### ISSUE TITLE
Connecting a machine to a network overrides default security group.

### ISSUE THREAD

**Issue by jwclark** [PSLCL/testing-framework#109]
QA has observed that connecting a machine to a network overwrites all of the existing security groups and replaces them with only the new network's security group.

Connecting a machine to a network should instead add the network's security group to all of the existing security groups currently attached to the machine instance.

Once the test is complete and the machine provider is cleaning up the instance in order to be re-used, all security groups should be removed from the instance except for the default security group configured in dtf.properties.

**Comment by Chad-Adams** [comment]
There were some solid errors in the code which have be fixed and promoted back to QA.  Other issues have popped which have kept QA from getting back to this point in the test automation.

Awaiting feedback from QA once they get back to the network binds steps in the sequence.

**Comment by amauchley** [comment]
So I followed Chad’s guidance on adding to the Runner node’s Role Policy to allow it to modify instance attributes, since he said that was what was missing for the security group issue that we were seeing.  When I made that change and executed a test, the security group that I told the runner to use via the dtf.properties file was never used but instead the default VPC security group was used.  So it appeared that the configured security group in the properties file wasn’t overridden this time, but not even used at all.  I’ve verified that my checkout of the testing-framework is up-to-date as well, and that indeed the new role policy was being used too.  So I’m confused as to why the framework is now doing an older previous issue that was resolved earlier.  You will see in  the log that STAF fails to ping the two machines, since the security group that should have been used had those ports opened, but was never attached.  These are the exceptions that happen after exceeding the limit of STAF Ping attempts:

com.pslcl.dtf.core.runner.resource.exception.FatalServerTimeoutException: resourceId: 4
at com.pslcl.dtf.resource.aws.ProgressiveDelay.retry(Unknown Source)
        at com.pslcl.dtf.resource.aws.ProgressiveDelay.retry(Unknown Source)
        at com.pslcl.dtf.resource.aws.instance.machine.MachineInstanceFuture.waitForStaf(Unknown Source)
        at com.pslcl.dtf.resource.aws.instance.machine.MachineInstanceFuture.createInstance(Unknown Source)
        at com.pslcl.dtf.resource.aws.instance.machine.MachineInstanceFuture.call(Unknown Source)
        at com.pslcl.dtf.resource.aws.instance.machine.MachineInstanceFuture.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

java.util.concurrent.ExecutionException: com.pslcl.dtf.core.runner.resource.exception.FatalServerTimeoutException: resourceId: 4
at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:192)
        at com.pslcl.dtf.runner.template.BindHandler.waitComplete(Unknown Source)
        at com.pslcl.dtf.runner.template.BindHandler.proceed(Unknown Source)
        at com.pslcl.dtf.runner.template.InstancedTemplate.runSteps(Unknown Source)
        at com.pslcl.dtf.runner.template.InstancedTemplate.<init>(Unknown Source)
        at com.pslcl.dtf.runner.template.TemplateProvider.getInstancedTemplate(Unknown Source)
        at com.pslcl.dtf.runner.process.RunEntryCore.testRun(Unknown Source)
        at com.pslcl.dtf.runner.process.Action$3.act(Unknown Source)
        at com.pslcl.dtf.runner.process.RunEntryTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

As I mentioned before the default VPC security group was the only security group attached to the nodes, I’m assuming that the intended behavior is to attach both security groups to the nodes (the one specified in the properties file, and the default VPC security group if the test requires a network). With that assumption, I was wondering if we even need to go down the road of attaching to security groups to the nodes that require networks.  Could we just simplify it, and only have the security group needing to be used as the one specified in the properties file?  It could just be an installation and usage choice for those using the framework that they decide all the ports, protocols, etc. needing to be opened for the type of testing they will do through the framework. If they plan on using networks in the generators, than they will need to open up the necessary ports, protocols on the specified security group, instead of us trying to attach the default security group combined with the specified one on the nodes. Just a thought, and I may be missing something else in the picture, but it seems to me that going this route wouldn’t harm any tests running on a standalone node while other nodes with tests having networks are running as well.
